### PR TITLE
Explore: Unified Node Graph Container

### LINKED ISSUE #69175: Explore: Unify Node Graph panel container
This task is a part of [Unify Panel containers epic.](https://github.com/grafana/grafana/issues/61563) 

TODO:
- [ ] Left-align panel title
- [ ] Use PanelChrome as container

### PR DESCRIPTION
**Pull request by harisrozajac**
NodeGraph panel now uses PanelChrome and it's consistent with other panels like Graph, Table, and Trace View. 

Since NodeGraph is collapsible when `withTraceView` is enabled, but the panel is closed by default, I've added this functionality to PanelChrome. Currently, PanelChrome supports collapsibility, but it doesn't have a way to set collapsed or expanded state of content on the initial render. It also doesn't support passing callbacks when toggling collapsed/expanded. To support this, I've added the following props: 
- `collapsed` that determines if the container is collapsed or expanded. 
- `onToggleCollapse` which is a callback that runs when toggling. In the case of NodeGraph this prop is used to run a function that captures user events and also to set the collapsed state. Logs follow the same [pattern](https://github.com/grafana/grafana/blob/0c072770ca6133226c9454a36c63747d59c1f97f/public/app/features/explore/Logs/Logs.tsx#L527C78-L527C121), so this will allow us to unify the Logs container too and make it use PanelChrome. 

Fixes: https://github.com/grafana/grafana/issues/69175

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] That it works as expected when `withTraceView` is enabled
- [ ] Let me know if the naming of these new props for PanelChrome makes sense. I've tried to come up with names that will not be confusing both to the user of the component and that will not make the PanelChrome's internal logic harder to read.

### CODE REVIEW

File: public/app/features/explore/NodeGraph/NodeGraphContainer.tsx
Comment (ifrost) [inline public/app/features/explore/NodeGraph/NodeGraphContainer.tsx:84]: This now shows `null` in the title when using test data:

Screenshot 2023-08-01 at 14 56 54

Comment (harisrozajac) [reply public/app/features/explore/NodeGraph/NodeGraphContainer.tsx]: Since `title` accepts only a string countWarning wouldn't display correctly if it had a value  because it's wrapped in a span. I've fixed this by passing `countWarning` to  PanelChrome's `titleItems` prop. I think it looks nice and it's good to have this message separate from the title. @grafana/observability-traces-and-profiling, any thoughts on this?

Comment (ifrost) [reply public/app/features/explore/NodeGraph/NodeGraphContainer.tsx]: Nice! I think just to make it consistent with what's currently in main we can align the message to the middle with:

```typescript
const getStyles = (theme: GrafanaTheme2) => ({
  warningText: css`
    display: flex;
    align-items: center;
    label: warningText;
    font-size: ${theme.typography.bodySmall.fontSize};
    color: ${theme.colors.text.secondary};
  `,
});
```

Screenshot 2023-08-02 at 08 53 41

File: packages/grafana-ui/src/components/PanelChrome/PanelChrome.tsx
Comment (Elfo404) [inline packages/grafana-ui/src/components/PanelChrome/PanelChrome.tsx]:
```suggestion
        onClick={toggle}
```

File: public/app/features/explore/NodeGraph/NodeGraphContainer.tsx
Comment (Elfo404) [inline public/app/features/explore/NodeGraph/NodeGraphContainer.tsx:91]:
```suggestion
      toggleCollapse={withTraceView ? toggled : undefined}
```

**Review by Elfo404 — COMMENTED** [review]
LGTM! just a couple of nits. 

would wait for owners to approve before merging

File: packages/grafana-ui/src/components/PanelChrome/PanelChrome.tsx
Comment (ivanortegaalba) [inline packages/grafana-ui/src/components/PanelChrome/PanelChrome.tsx:77]: Why don't we just pass if it is collapsed?
```
collapsed={true|false}
```
So we can make the panel a controlled state.

```jsx
function MyPage(){
  const [collapsed, setCollapsed]
  return (
    <PanelChrome
      collapsible
      collapsed={collapsed}
      onToggleCollapse={(collapsed)=>setCollapsed(collapsed)}
  )
}
```
I have to say, it is weird to have `collapsible` and `collapsed` as part of the same component API.
What do we do if we pass `collapsed` and `onToggleCollapse` but not `collapsible`?

How it should work:
1. If we pass `collapsed` it becomes collapsible but it doesn't change the state when clicking because the component becomes controlled. Like an input when you pass a `value`
2. If we pass `onToggleCollapse` but not collapsed, it doesn't become collapsible
3. If we pass `collapsed` and `onToggleCollapse` the component becomes collapsible, so the prop `collapsible` is not required
4. If we pass only `collapsible`, the component is uncontrolled, so it collapse by itself, with the internal state. If we pass the callback, the callback is called, but the state is internal until we don't pass `collapsible` as prop.

Is it clear? Sorry, quite hard to put everything on the table 😄

**Review by ivanortegaalba — CHANGES_REQUESTED** [review]
I requested some changes in the API, please let me know what do you think 😄

File: packages/grafana-ui/src/components/PanelChrome/PanelChrome.tsx
Comment (harisrozajac) [reply packages/grafana-ui/src/components/PanelChrome/PanelChrome.tsx:77]: @ivanortegaalba thank you for your feedback on this. 

I understand what you are saying and here some clarifying comments and questions:

> What do we do if we pass `collapsed` and `onToggleCollapse` but not `collapsible`?

This wouldn't be possible because of how the `Collapsible` interface is defined. `collapsible` always has to be passed if `defaultExpanded` or `onToggleCollapse` are to be used.

> 1. If we pass collapsed it becomes collapsible but it doesn't change the state when clicking because the component becomes controlled. Like an input when you pass a value
> 2. If we pass `onToggleCollapse` but not collapsed, it doesn't become collapsible

Why would we pass `onToggleCollapse` or 'collapsed` just by themselves if the component is not collapsible? 

In my mind, I think it's more readable to first declare if component is collapsible, and then pass other props related to that property, like if the component should be collapsed on first render and then pass any callbacks on toggling.

Comment (ivanortegaalba) [reply packages/grafana-ui/src/components/PanelChrome/PanelChrome.tsx:77]: Thanks for the answer @harisrozajac 

> This wouldn't be possible because of how the Collapsible interface is defined. collapsible always has to be passed if defaultExpanded or onToggleCollapse are to be used.

Yep, I understand this would be a breaking change, because `collapsible` is already released.

> Why would we pass onToggleCollapse or collapsed just by themselves if the component is not collapsible?

When you pass them, you're implicitly declaring the component as collapsible. In the same way when you declare `onClick` you are making the component clickable but you don't add a property called `clickable`.

> In my mind, I think it's more readable to first declare if component is collapsible, and then pass other props related to that property, like if the component should be collapsed on first render and then pass any callbacks on toggling.

My suggestion is not about readability, is more about standardization. When you have an input text or any other component that can receive a state from outside, you don't rely on an internal state, this is what is known as [a controlled component](https://legacy.reactjs.org/docs/forms.html#controlled-components).

With this API, you cannot make the component controlled, since the component has an internal state that you cannot force or change from outsize. You would need a full reset of the component to change the state from outside. For example, if I want to have a keyboard shortcut that toggles all the panels, how would you implement it? You must apply the logic inside the component. Or if you want to open the panel when another element is clicked, you cannot make it happen. 

This is why I miss the possibility to make the component controlled, and you can only do that by allowing `collapsed` property. But if you allow `collapsed`, `defaultExpanded` doesn't make sense anymore.

So, my new suggestion, to combine your approach with the controlled component is:
```jsx
function MyPage(){
  // By default will be open
  const [isCollapsed, setCollapsed] = useState(true)

  return (
    <PanelChrome
      {/* This must be provided to make the component collapsible */}
      collapsible
      {/* This will expand or collapse no matter the internal state. If it is not provided, the component will be expanded by default, and it will have the toggle behavior as usual */}
      collapsed={isCollapsed}
      {/* This will receive the new state, toggling the previous one */}
      onToggleCollapse={(isCollapsed)=>setCollapsed(isCollapsed)}
  ) />
}
```

What do you think?

Comment (harisrozajac) [reply packages/grafana-ui/src/components/PanelChrome/PanelChrome.tsx:77]: @ivanortegaalba thank you for the clarification. It makes sense and I'll make the adjustments 😄

File: packages/grafana-ui/src/components/PanelChrome/PanelChrome.test.tsx
Comment (ivanortegaalba) [inline packages/grafana-ui/src/components/PanelChrome/PanelChrome.test.tsx:42]: I think we should have two cases covered here: Controlled and uncontrolled

```jsx
// Controlled
<PanelChrome collapsible collapsed={collapsed} onToggleCollapse={toggleCollapsed} />

// Uncontrolled
<PanelChrome collapsible />
```
And both should collapse properly

File: public/app/features/explore/NodeGraph/NodeGraphContainer.tsx
Comment (ivanortegaalba) [inline public/app/features/explore/NodeGraph/NodeGraphContainer.tsx:92]: Do you need to control the state of the component? Can you leave the component uncontrolled?

**Review by ivanortegaalba — COMMENTED** [review]
I left some small comments about missing test cases. The code looks good to me 😄

File: public/app/features/explore/NodeGraph/NodeGraphContainer.tsx
Comment (harisrozajac) [reply public/app/features/explore/NodeGraph/NodeGraphContainer.tsx:92]: I do need to control it because when trace view is enabled, the component needs to be collapsed by default, while uncontrolled one would be expanded by default.

**Review by mckn — COMMENTED** [review]
LGTM! But wait for code owners to approval before merge.

**Review by joey-grafana — APPROVED** [review]

**Review by ivanortegaalba — APPROVED** [review]
Thanks for applying the feedback @harisrozajac! LGTM!

**Pull request merged**
